File: LeetCode/Medium/0797_All_Paths_From_Source_to_Target/Solution.py
## 2021-12-14 (recursive backtracking DFS)
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        possible_paths = []
        self.find_paths(graph, 0, possible_paths, [])
        return possible_paths
    
    def find_paths(self, graph, position, possible_paths, current_path):
        current_path.append(position)
        
        if position == len(graph)-1:
            possible_paths.append(current_path.copy())
        else:
            for neighbour in graph[position]:
                self.find_paths(graph, neighbour, possible_paths, current_path)
        
        current_path.pop()

## 2023-02-16 (recursive backtracking
# class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        """
        input: a DAG
        output: all possible paths from node 0 to node n-1 (any order)

        notes:
        - simple backtracking - just processing every single possibility
        - we'd probably approach this recursively
            - function with a list to track path, a global with our answers
            - recurse down every neighbour until we hit n-1, or until there
              are no other neighbours (acyclic)
            - if we hit n-1, add to our answer list and go backwards
                - pop off our current node to make room for the other paths

        analysis:
        - time = O(e) where e=number of edges in the graph
        - space = O(v) where v=number of vertices (if we don't include answer output)
        """
        answers = []
        
        def findAllPaths(node, graph, path):
            nonlocal answers
            path.append(node)
            
            if node == len(graph)-1:
                answers.append(path[:])
            else:
                for neighbour in graph[node]:
                    findAllPaths(neighbour, graph, path)
            path.pop()
        
        findAllPaths(0, graph, [])
        return answers

File: LeetCode/Medium/0797_All_Paths_From_Source_to_Target/Solution.cpp
// 2022-12-29 (iterative DFS)
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        /*
            input: an adjacency list of a directed acyclic graph
            output: all possible paths from node 0 to node n-1

            notes:
            - acyclic, so don't need to worry about cycles
            - most likely just backtracking DFS, starting from node 0
            - adjacency list is made for us! yay!
            
            approach:
            - keep a stack of pair<current node, current path> and perform iterative DFS
            - store our current node into the current path up to the node
            - every time we reach the end node, store the path into an output vector
            - otherwise, just push the neighbour + path into our stack and continue
        */
        int n = graph.size();
        vector<vector<int>> output;
        deque<pair<int, vector<int>>> stack = {{0, {}}};
        while (!stack.empty()) {
            auto [node, path] = stack.back();
            stack.pop_back();
            path.push_back(node);
            if (node == n-1) {
                output.push_back(path);
            } else {
                for (int neighbour : graph[node]) {
                    stack.push_back({neighbour, path});
                }
            }
        }
        return output;
    }
};